### PR TITLE
Add support for Go 1.7

### PR DESCRIPTION
**Pull request by HelloGrayson**
Let's get ahead of 1.7 before it's released.

### CODE REVIEW

**Comment by HelloGrayson** [comment]
Hmmm, `no buildable Go source files`... perhaps there's a different linter for 1.7?

```
go get -u -f github.com/golang/lint/golint
package github.com/golang/lint/golint
    imports github.com/golang/lint
    imports go/types: unrecognized import path "go/types"
package github.com/golang/lint/golint
    imports github.com/golang/lint
    imports golang.org/x/tools/go/gcimporter15
    imports golang.org/x/tools/go/gcimporter15
    imports golang.org/x/tools/go/gcimporter15: no buildable Go source files in /home/travis/gopath/src/golang.org/x/tools/go/gcimporter15
```

**Comment by prashantv** [comment]
All the files in `gcimporter15` have build tags that restrict what versions of Go they'll build with.
https://github.com/golang/tools/tree/master/go/gcimporter15

There ends up being no files that can be built in Go 1.7. It's possible that the linter just hasn't been updated to work with Go 1.7 yet. I'd suggest only installing golint and running it on go 1.5 or go 1.6 for now.

**Comment by HelloGrayson** [comment]
@abhinav you've done matrix stuff like this before with Travis - is it fairly straightforward?

**Comment by abhinav** [comment]
Yeah, you can replace the `go:` section with something like,

``` yaml
matrix:
  include:
    - go: 1.5
      env: LINT=0
    - go: 1.6
      env: LINT=1
    - go: 1.7
      env: LINT=0
```

And change your script to lint only if `$LINT == 1`.

**Comment by prashantv** [comment]
We do ugly things in tchannel-go's Makefile to detect the Go version and decide whether to run lint:
https://github.com/uber/tchannel-go/blob/dev/Makefile#L6

**Comment by HelloGrayson** [comment]
Ok, fixed the first issue. Now I'm getting this:

```
0.38s$ make test_ci
DEPRECATED Action signature.  Must be `cli.ActionFunc`.  This is an error in the application.  Please contact the distributor of this application if this is not you.  See https://github.com/urfave/cli/blob/master/CHANGELOG.md#deprecated-cli-app-action-signature
./scripts/cover.sh github.com/crossdock/crossdock-go
testify.go:27:2: cannot find package "github.com/stretchr/testify/assert" in any of:
    /home/travis/.gimme/versions/go1.4.1.linux.amd64/src/github.com/stretchr/testify/assert (from $GOROOT)
    /home/travis/gopath/src/github.com/stretchr/testify/assert (from $GOPATH)
call.go:30:2: cannot find package "github.com/stretchr/testify/require" in any of:
    /home/travis/.gimme/versions/go1.4.1.linux.amd64/src/github.com/stretchr/testify/require (from $GOROOT)
    /home/travis/gopath/src/github.com/stretchr/testify/require (from $GOPATH)
call.go:31:2: cannot find package "golang.org/x/net/context" in any of:
    /home/travis/.gimme/versions/go1.4.1.linux.amd64/src/golang.org/x/net/context (from $GOROOT)
    /home/travis/gopath/src/golang.org/x/net/context (from $GOPATH)
call.go:32:2: cannot find package "golang.org/x/net/context/ctxhttp" in any of:
    /home/travis/.gimme/versions/go1.4.1.linux.amd64/src/golang.org/x/net/context/ctxhttp (from $GOROOT)
    /home/travis/gopath/src/golang.org/x/net/context/ctxhttp (from $GOPATH)
make: *** [test_ci] Error 1
```

Don't see anything obvious... it looks like `make install_ci` executed fine above...

**Comment by prashantv** [comment]
I found it a little odd that the go1.7 build failed with an error that mentions go1.4.1:

```
testify.go:27:2: cannot find package "github.com/stretchr/testify/assert" in any of:
    /home/travis/.gimme/versions/go1.4.1.linux.amd64/src/github.com/stretchr/testify/assert (from $GOROOT)
    /home/travis/gopath/src/github.com/stretchr/testify/assert (from $GOPATH)
```

Turns out the gimme is failing:

```
$ eval "$(gimme 1.7)"
I don't have any idea what to do with '1.7'.
```

I wonder if all the go1.7 builds are silently running with 1.4 since Travis isn't failing the build at that point immediately (seems like a pretty serious issue..)

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
 
-go:
-    - 1.5
-    - 1.6
+matrix:
+    include:
+        - go: 1.5
+          env: LINT=1
+        - go: 1.6
+          env: LINT=1
```
Comment (abhinav) [inline .travis.yml]: You don't want to lint in both 1.5 and 1.6. There are some conflicting rules between the two versions. In thriftrw and yarpc go, we just lint for 1.6.

**Comment by HelloGrayson** [comment]
@prashantv yeah, looks like 1.7 isn't avail in gimme yet:

```
$ brew install gimme

$ eval "$(GIMME_GO_VERSION=1.7 gimme)"
I don't have any idea what to do with '1.7'.
  (using type 'auto')

$ eval "$(GIMME_GO_VERSION=1.6 gimme)"
go version go1.6 darwin/amd64

$ eval "$(GIMME_GO_VERSION=1.5 gimme)"
go version go1.5 darwin/amd64

$ eval "$(GIMME_GO_VERSION=1.4 gimme)"
go version go1.4 darwin/amd64

$ gimme list
1.4 <= current
1.5
1.6
```

Then... what are we doing in the other repos...

**Comment by HelloGrayson** [comment]
Check out [yarpc-go's 1.7 test](https://travis-ci.org/yarpc/yarpc-go/jobs/142840909):

```
19.09s$ eval "$(gimme 1.7)"
I don't have any idea what to do with '1.7'.
  (using type 'auto')
$ export GOPATH=$HOME/gopath
$ export PATH=$HOME/gopath/bin:$PATH
$ mkdir -p $HOME/gopath/src/github.com/yarpc/yarpc-go
$ rsync -az ${TRAVIS_BUILD_DIR}/ $HOME/gopath/src/github.com/yarpc/yarpc-go/
$ export TRAVIS_BUILD_DIR=$HOME/gopath/src/github.com/yarpc/yarpc-go
$ cd $HOME/gopath/src/github.com/yarpc/yarpc-go
$ gimme version
v0.2.3
$ go version
go version go1.5.1 linux/amd64
```

It's using `go1.5.1` 

:|

**Comment by HelloGrayson** [comment]
Because of https://github.com/travis-ci/travis-ci/issues/6299, I've updated Travis to use `1.7rc1` and `tip`.

**Pull request merged**